031-tls-encrypt-the-kafka-connect-rest-api.md

```diff
@@ -0,0 +1,172 @@
+
+## Current situation
+
+Currently, instances of Kafka Connect that are deployed by the Strimzi operators are configured with the default REST API endpoint settings.
+This means that the Kafka Connect REST API endpoint uses HTTP on port 8083 and that the Strimzi `KafkaConnect`, `KafkaConnector` and `KafkaMirrorMaker2` operators make unsecured REST client calls based on this default configuration.
+
+The default network policies created by the Strimzi operators restrict incoming REST API calls to only allow access from the operator pod.
+Users can define further network policies to override the default policy and allow wider access.
+
+## Motivation
+
+There is currently no way to TLS encrypt the Kafka Connect REST API, leaving the Connect REST listener as perhaps the last key endpoint that cannot currently be encrypted using Strimzi.
+
+## Proposal
+
+This proposal adds REST api configuration options to the `KafkaMirrorMaker2` and `KafkaConnect` CRDs.
+
+A new field is added to the `KafkaConnect` and `KafkaMirrorMaker2` specs named `restListeners`, which allows users to configure the Kafka Connect configuration to have:
+  - a single HTTP listener on port 8083
+  - a single encrypted HTTPS listener on port 8443
+  - both an HTTP listener on port 8083 and an encrypted HTTPS listener on port 8443
+  
+This enables users to access the REST API with TLS if required.
+The `KafkaConnector`, `KafkaConnect` and `KafkaMirrorMaker2` operators will use the TLS encrypted listener when it is enabled, even if an unencrypted listener is present.
+
+The `restListeners` field is designed for potential future extension, and is a list of items that contains a `tls` field and fields for providing certificates if required `certChainAndKey`, `trustedCertificates`
+
+The `tls` field is a boolean, if it is `true` the listener is HTTPS with port 8443, if it is `false` the listener is HTTP with port 8083. If the `restListeners` list is empty, or not specified at all, the `KafkaConnect` and `KafkaMirrorMaker2` 
+runtimes will be created with a single unencrypted listener on 8083 to match the existing behaviour.
+
+The `certChainAndKey` field will be required if `tls` is `true`. The `trustedCertificates` will be required if `tls` is `true` and the provided 
+certificate is not signed by a CA root certificate that is trusted by the Java runtime by default (i.e. signed by a well known CA like
+Verisign or Let's Encrypt).
+
+For example, the following CR will enable the HTTP listener on port 8083 and the HTTPS listener on port 8443:
+```
+apiVersion: kafka.strimzi.io/v1beta1
+kind: KafkaConnect
+spec:
+  restListeners:
+  - tls: true
+    certChainAndKey:
+      secretName: my-secret
+      certificateChain: my-certificate.crt
+      key: my-key.key
+    trustedCertificates:
+      secretName: my-secret
+      certificate: ca.crt
```

Comment:
Should this be a list to allow multiple trusted certificates to be added?

Reply:
The intention was that the user would specify the root certificate here, but if they need to update the certificates then I think having the option to add multiple trusted certificates would help with that so I'll update this to be an array.